vm-allocator/src/system.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2018 The Chromium OS Authors. All rights reserved.
+// Copyright © 2019 Intel Corporation
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+use vm_memory::{GuestAddress, GuestUsize};
+
+use crate::address::AddressAllocator;
+use crate::id::IdAllocator;
+
+use libc::{sysconf, _SC_PAGESIZE};
+use std::fmt::{self, Display};
+use std::result;
+use std::sync::{Arc, Mutex};
+
+/// Errors associated with system resources allocation.
+#[derive(Debug)]
+pub enum Error {
+    /// Address is none.
+    NoneAddress,
+    /// Address allocation failed.
+    AddressAllocate(crate::address::Error),
+    /// Id allocation failed.
+    IdAllocate(crate::id::Error),
+}
+
+impl Display for Error {
+    // This trait requires `fmt` with this exact signature.
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            NoneAddress => write!(f, "The address being allocated is null"),
```

Comment:
`null` is improperly said since null doesn't really exist in its pure form in Rust. Also, the emphasize should be on the operation that failed (i.e. the allocation) and not on the object being used. I would rephrase this to: "Cannot allocate address. The I/O address space is undefined." This would require splitting the errors per address space as I suggested in my previous comment.

Reply:
Got it. So let's change the error as
UndefinedAddress => write!(f, "Cannot allocate address. The address being allocated is undefined"),
This kind of error would only happen during PIO allocation.

Reply:
`NoneAddress` is returned when `io_adress_space` is not initialized so I don't think that `UndefinedAddress` is an accurate name.

We can have a generic error like `UninitializedAllocator(&'static str)` which we can use for all allocators that are optional. When an `allocate_` function is called for an allocator that was not initialized, we can return `UninitializedAllocator("allocator_name")`.

```rust
pub enum Error {
    /// Allocator is not initialized.
    UninitializedAllocator(&'static str),
    /// MMIO allocation failed.
    AllocateMmio(crate::address::Error),
    /// I/O allocation failed.
    AllocateIO(crate::address::Error),
    /// IRQ allocation failed.
    AllocateIrq(crate::id::Error),
   /// device_id allocation failed.
    AllocateDeviceId(crate::id::Error),
}

impl Display for Error {
    // This trait requires `fmt` with this exact signature.
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use self::Error::*;

        match self {
            // When implementing display we can print the name of the uninitialized allocator so it makes it easier to debug. 
            UninitializedAllocator(allocator_name) => write!(f, "The allocator {} is not initialized.", allocator_name),
            ....  
        }
    }
}
```

Let's assume that both I/O Allocator and IRQ Allocator are optional. In this case both `allocate_io` and `allocate_irq` can fail with the same error `UninitializedAllocator`, but we can differentiate them because the error contains the name of the allocator that failed.

```rust
    /// Reserves a section of `size` bytes of IO address space.
    pub fn allocate_io_addresses(
        &mut self,
        address: GuestAddress,
        size: GuestUsize,
    ) -> Result<GuestAddress> {
        if let Some(io_allocator) = self
            .io_address_space
            .lock()
            .expect("failed to acquire lock")
            .as_mut()
        {
            ....
        } else {
            Err(Error::UninitializedAllocator("io_allocator"))
        }
    }

....

    pub fn allocate_irq(&mut self, irq: Option<u32>) -> Result<u32> {
         if let Some(irq_allocator) = self.irq
            .lock()
            .expect("failed to acquire lock") {
             ......
         } else {
            Err(Error::UninitializedAllocator("irq_allocator"))
         }
    }

```